### PR TITLE
Pass missing arguments to `SeamlessM4Tv2ConformerEncoderLayer.forward()` when gradient checkpointing is enabled

### LINKED ISSUE #31028: `SeamlessM4Tv2ConformerEncoder` does not behaves as expected if gradient checkpointing is enabled
### System Info

- `transformers` version: 4.42.0.dev0
- Platform: Linux-5.4.0-172-generic-x86_64-with-glibc2.17
- Python version: 3.8.19
- Huggingface_hub version: 0.23.1
- Safetensors version: 0.4.3
- Accelerate version: 0.30.1
- Accelerate config:    not found
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- Tensorflow version (GPU?): 2.13.1 (True)
- Flax version (CPU?/GPU?/TPU?): 0.7.0 (cpu)
- Jax version: 0.4.13
- JaxLib version: 0.4.13
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@ArthurZucker @sanchit

Proposed fix:
```python
class SeamlessM4Tv2ConformerEncoder(...):
    [...]
    def forward(...):
        [...]
            if not skip_the_layer or deepspeed_zero3_is_enabled:
                # under deepspeed zero3 all gpus must run in sync
                if self.gradient_checkpointing and self.training:
                    layer_outputs = self._gradient_checkpointing_func(
                        layer.__call__,
                        hidden_states,
                        attention_mask,
                        output_attentions,    # <---------- Add this parameter
                        conv_attention_mask,  # <---------- Add this parameter         
                    )
                else:
                    layer_outputs = layer(
                        hidden_states,
                        attention_mask=attention_mask,
                        output_attentions=output_attentions,
                        conv_attention_mask=conv_attention_mask,
                    )
                hidden_states = layer_outputs[0]
        [...]
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

1. Train a model that has `transformers.models.seamless_m4t_v2.modeling_seamless_m4t_v2.SeamlessM4Tv2ConformerEncoder` as a submodule
2. (enable gradient checkpointing while training)
3. When calling `SeamlessM4Tv2ConformerEncoder.forward()`, pass `output_attentions=True` and `return_dict=True`. For example:
    ```python
    encoder: SeamlessM4Tv2ConformerEncoder = ...
    output = encoder(..., output_attentions=True, return_dict=True)
    ```

### Expected behavior

`output.attentions` is a tuple of not-None tensors, one per encoder layer. Instead, the actual behavior is that `output.attentions = (None, None, ..., None)`.

### PR DESCRIPTION
**Pull request by anferico**
# What does this PR do?
Fixes a bug for which if gradient checkpointing is enabled, `SeamlessM4Tv2ConformerEncoderLayer.forward()` is called with some missing arguments.

Fixes #31028 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ylacombe

### CODE REVIEW

**Review by ylacombe — APPROVED** [review]
LGTM ! Thanks for fixing this! 
Could you also verify that this doesn't happen in the v1 of `SeamlessM4T` ?

**Comment by anferico** [comment]
@ylacombe turns out the same problem existed in v1 too 🤓 Just pushed a new commit fixing that

**Comment by ylacombe** [comment]
Great! 
@amyeroberts, the PR is ready for a review. Thanks in advance!

**Comment by amyeroberts** [comment]
@anferico Thanks for adding this fix! 

This should be caught in tests. Could you check the test suit for this model to make sure the forward pass with gradient checkpoint is tested?

**Comment by anferico** [comment]
@amyeroberts The gradient checkpointing-related tests are currently being skipped for both SeamlessM4T [v1](https://github.com/huggingface/transformers/blob/72fb02c47dbbe1999ae105319f24631cad6e2e00/tests/models/seamless_m4t/test_modeling_seamless_m4t.py#L494) and [v2](https://github.com/huggingface/transformers/blob/72fb02c47dbbe1999ae105319f24631cad6e2e00/tests/models/seamless_m4t_v2/test_modeling_seamless_m4t_v2.py#L510).

Aside from that, this issue kinda calls for more robust tests, because the [current test suite](https://github.com/huggingface/transformers/blob/72fb02c47dbbe1999ae105319f24631cad6e2e00/tests/test_modeling_common.py#L777) for gradient checkpointing only checks if gradients are computed for each parameter, but it could never have spotted the edge case where the `model.forward()` method behaves differently depending on whether GC is enabled or not. I understand it might be a big effort, but ideally we would require every function that is affected by GC be tested twice: once with GC enabled and once with GC disabled, making sure that the outputs are exactly the same given certain inputs. What do you think?

**Review by amyeroberts — APPROVED** [review]

**Comment by amyeroberts** [comment]
@anferico OK, sounds like the work is a bit more involved. Could you open an issue detailing that the GC tests should be updated? This way we keep track of the work and we can merge this without being blocked

**Comment by anferico** [comment]
@amyeroberts done! #32063

**Comment by amyeroberts** [comment]
@anferico Awesome - thank you!
